gcp_variant_transforms/beam_io/vcfio.py

```diff
@@ -207,6 +210,124 @@ def __repr__(self):
+    encoded_calls = self._encode_variant_calls(variant, format_keys)
+
+    columns = [
+        variant.reference_name,
+        variant.start if variant.start is None else variant.start + 1,
+        ';'.join(variant.names),
+        variant.reference_bases,
+        ','.join(variant.alternate_bases),
+        variant.quality,
+        ';'.join(variant.filters),
+        encoded_info,
+        ':'.join(format_keys),
+    ]
+    if encoded_calls:
+      columns.append(encoded_calls)
+    columns = [self._encode_value(c) for c in columns]
+
+    return '\t'.join(columns) + '\n'
+
+  def _encode_value(self, value):
+    """Encodes a single :class:`Variant` column value for a VCF file line."""
+    if not value and value != 0:
+      return MISSING_FIELD_VALUE
+    elif isinstance(value, list):
+      return ','.join([self._encode_value(x) for x in value])
+    return str(value)
+
+  def _encode_variant_info(self, variant):
+    """Encodes the info of a :class:`Variant` for a VCF file line."""
+    encoded_infos = []
+    # Set END in info if it doesn't match start+len(reference_bases). This is
+    # usually the case for non-variant regions.
+    if (variant.start is not None
+        and variant.reference_bases
+        and variant.end
+        and variant.start + len(variant.reference_bases) != variant.end):
+      encoded_infos.append('END=%d' % variant.end)
+    # Set all other fields of info.
+    for k, v in variant.info.iteritems():
+      if v.data is True:
+        encoded_infos.append(k)
+      else:
+        encoded_infos.append('%s=%s' % (k, self._encode_value(v.data)))
+    return ';'.join(encoded_infos)
+
+  def _get_variant_format_keys(self, variant):
+    """Gets the format keys of a :class:`Variant`."""
+    if not variant.calls:
+      return []
+
+    format_keys = [GENOTYPE_FORMAT_KEY]
+    for call in variant.calls:
+      # If any calls have a set phaseset that is not `DEFAULT_PHASESET_VALUE`,
+      # the key will be added to the format field.
+      if self._is_alternate_phaseset(call.phaseset):
+        format_keys.append(PHASESET_FORMAT_KEY)
+      format_keys.extend([k for k in call.info])
+
+    # Sort all keys and remove duplicates after GENOTYPE_FORMAT_KEY
+    format_keys[1:] = sorted(list(set(format_keys[1:])))
+
+    return format_keys
+
+  def _encode_variant_calls(self, variant, format_keys):
+    """Encodes the calls of a :class:`Variant` in a VCF line."""
+    # Ensure that genotype is always the first key in format_keys
+    assert not format_keys or format_keys[0] == GENOTYPE_FORMAT_KEY
+    encoded_calls = []
+    for call in variant.calls:
+      encoded_call_info = [self._encode_genotype(call.genotype,
+                                                 call.phaseset)]
+      for key in format_keys[1:]:
+        if key == PHASESET_FORMAT_KEY:
+          encoded_call_info.append(self._encode_phaseset(call.phaseset))
+        else:
+          encoded_call_info.append(self._encode_call_info_value(call.info,
```

Comment:
nit: I think it's more readable to break it after .append and start a new line at self._encode...

Reply:
Good call!